qt/applications/workbench/workbench/app/mainwindow.py

```diff
@@ -361,6 +367,40 @@ def save_file(self):
     def open_manage_directories(self):
         ManageUserDirectories(self).exec_()
 
+    def readSettings(self, settings):
+        qapp = QApplication.instance()
+        qapp.setAttribute(Qt.AA_UseHighDpiPixmaps)
+        if hasattr(Qt, 'AA_EnableHighDpiScaling'):
+            qapp.setAttribute(Qt.AA_EnableHighDpiScaling, settings.get('main/high_dpi_scaling'))
+
+        # restore window geometry
+        window_size = settings.get('main/window/size')
+        if not isinstance(window_size, QSize):
+            window_size = QSize(*window_size)
+        self.resize(window_size)
+        window_pos = settings.get('main/window/position')
+        if not isinstance(window_pos, QPoint):
+            window_pos = QPoint(*window_pos)
+        self.move(window_pos)
+
+        # restore window state
+        windowstate = Qt.WindowNoState
+        if settings.has('main/window/state'): # preferred
+            settings.get('main/window/state')
+        elif settings.get('main/window/is_maximized'):
+            windowstate = Qt.WindowMaximized
+        elif settings.get('main/window/is_fullscreen'):
+            windowstate = Qt.WindowFullScreen
+        self.setWindowState(windowstate)
+
+    def writeSettings(self, settings):
+        settings.set('main/window/size', self.size())
+        settings.set('main/window/position', self.pos())
+
+        # remove keys that come from the defaults
+        settings.set('main/window/state', self.saveState())
+        settings.remove('main/window/is_maximized')
```

Comment:
Can we avoid making a special case for these and inspect `config.DEFAULTS` in some way?